### DISCUSSION TITLE
Add files to serverless function in postbuild

### DISCUSSION THREAD

**Discussion by tonprince** [vercel/vercel#6461] [Help]
My serverless function contains a search-index logic and reads binary index files from the filesystem. To make that code working I included the index files via 'includeFiles' in vercel.json. That is working fine so far.

As the index files often change, I would like to automate the creation of the index files directly within a postbuild step.
Is there a way to write files to the serverless function folder during the build process?

I already tried it without success, and the serverless function always complains about missing files during execution.

Does anyone know how the serverless function is packed and deployed, perhaps that would help to understand and find a solution for this issue.

**Comment by barelyhuman** [comment]
A few questions regarding the same 
1. Git connected deploys? 
2.  CLI deploys?


Based on the above you can try a few things. 
I've got a similar thing but instead of post build, I have a script that generates the needed assets right before the actual build. If it's possible for you to do that, you can go ahead and do that. 
You probably already know this but the setup for the git connected version can be done on App > Settings > General > Build Command

and you just add you bash/node script to be run after build using something like, `npm run build:api; node generate-indexes.js`

I'm don't use the cli much so I'm not sure of handling it there, I'll let you know after I check. 

I hope the above helps.

**Reply by tonprince** [reply]
Thanks. I already tried to put a pre-build task via Build Command in Vercel, but it seams the build process for serverless functions is a bit different so my index files are not picked up and finally not included in the function.

I could solve this issue by implementing a Github action which runs the build index script, pushes the changes back to main branch and then triggers a Vercel build. In this case the index files are included in the serverless function.

**Reply by barelyhuman** [reply]
Hmm, any chance you can create a similar repo with a minimal reproduction of the same? 
I'd like to fork it and try the same ?

**Reply by tonprince** [reply]
Sure, I just created a minimal Git repo here:
https://github.com/tonprince/serverless-search-index

After fork and linking to Vercel, you can just make some local changes and commit them to the main branch. This will trigger the GitHub action which will create a new index based on your commit message. After successfully deployment on Vercel, you can query the endpoint under `api/index?query='YOUR COMMIT MESSAGE'`. 

Here some remarks:

- The index files are copied to the tmp folder on the serverless function server to bypass a readonly filesystem issue. Even if we only query the index, it is required to create a lock which is only possible in the temp folder.
- The property `github.enabled` in `vercel.json` is initially set to `false`, to prevent that Vercel will start building without the required index files. The GitHub action script will set the flag to true and commit the changes together with the index files to the `main` branch. Afterwards, the flag is reset to `false`.
- For the first commit and test, I recommend to delete the existing index files in the repo under `api/comments` to have a clean setup.

Finally you should see the query result when calling your endpoint:

![image](https://user-images.githubusercontent.com/6906225/125323783-645fc000-e369-11eb-9c6f-746e95ab5685.png)

**Reply by barelyhuman** [reply]
While I realised that the arch would create a new instance on each invoke, I wasn't expecting it to have only the used files for the api and public folder.

So your best bet is to use a precommit hook to generate the files as needed. Don't really need actions for it if the use case can generate files during push. 

If the actual usecase needs to generate the index in a different manner, then the actions method is your way to go. 

Finally, 
1. You need the files to be a part of the repo before the deploy starts 
2. You don't need to have actions handle it for you 

The reason is that includeFiles works for existing files and not build time generated files, post build, the actual actions are copied into executor instances that are launched during invocation and thus have only /tmp to their access and each invoke has a different /tmp instance

hope that explains the issue?

**Reply by tonprince** [reply]
Thanks

**Reply by barelyhuman** [reply]
You could also try the vercel-build step to do the same though. 
if you haven’t already tried it out. 

You can find info about it here. 
https://vercel.com/docs/runtimes?query=Custom%20build%20step#advanced-usage/advanced-node-js-usage/custom-build-step-for-node-js

**Reply by tonprince** [reply]
I finally added the build index script within a vercel-build step, that works fine and does not require to commit the index files to repo.

**Reply by barelyhuman** [reply]
I’m glad it worked out.

**Reply by tonprince** [reply]
Do you know how to trigger a vercel build with deploy hook if the flag github.enabled is set to false?

**Reply by barelyhuman** [reply]
There’s 2 apis. 
one to upload the files 
2nd to run the deployment for the uploaded files. 

I’ve not tried them but let me see what I can provide as a reference.

**Reply by barelyhuman** [reply]
@tonprince You can go ahead and use the following references . 
https://vercel.com/docs/api#endpoints/deployments/create-a-new-deployment
for creating a deployment 
https://vercel.com/docs/api#endpoints/deployments/upload-deployment-files
for uploading the deployment files 

as for the bearer token needed for these, you can find that in your account => settings => token => create token 

You could just setup vercel cli in the CI to do it for you with the same token if you don't plan to use the deploy apis. 

Also , if possible select the answer as answered so others can refer to it for solutions later on.
Thanks

**Reply by tonprince** [reply]
Hi
is there a way to start a post build script for a serverless function when everything is deployed and have access to tmp folder?

I need to run a script to setup the db but when doing so in a HTTP request I run into a timeout.

**Reply by barelyhuman** [reply]
I haven't tried that but based on just understanding of the product it's not something you want to do. 
If you are doing this to create a temporary database you are better of using something like https://upstash.com/ for things like these

**Reply by tonprince** [reply]
Unfortunately I am stick to search-index npm library.
